File: frontend/src/api/v3/health.ts
/**
 * V3 Health API Functions - MDP Platform V3.1
 * API client for index health monitoring and observability.
 */
import v3Client from './client';

// ==========================================
// Types - Metrics
// ==========================================

export interface IJobMetrics {
  pk_collisions: number;
  ai_latency_avg_ms: number;
  ai_inference_count: number;
  ai_low_confidence_count: number;
  vector_dim_mismatch: number;
  corrupt_media_files: number;
  transform_errors: number;
}

// ==========================================
// Types - Object Health
// ==========================================

export type HealthStatus = 'HEALTHY' | 'DEGRADED' | 'FAILED';

export interface IObjectHealthSummary {
  object_def_id: string;
  object_name: string | null;
  status: HealthStatus;
  last_run_time: string | null;
  lag_seconds: number | null;
  rows_processed: number;
  rows_indexed: number;
  success_rate: number;
  metrics: Partial<IJobMetrics>;
}

export interface ISystemHealthSummary {
  total_objects: number;
  healthy_count: number;
  degraded_count: number;
  failed_count: number;
  overall_success_rate: number;
  avg_ai_latency_ms: number | null;
  total_corrupt_files: number;
  objects: IObjectHealthSummary[];
}

// ==========================================
// Types - Job Run
// ==========================================

export type JobStatus = 'SUCCESS' | 'PARTIAL_SUCCESS' | 'FAILED';

export interface IIndexJobRun {
  id: string;
  mapping_id: string;
  object_def_id: string;
  start_time: string;
  end_time: string | null;
  status: JobStatus;
  rows_processed: number;
  rows_indexed: number;
  metrics_json: Partial<IJobMetrics>;
  created_at: string | null;
}

export interface IJobHistoryResponse {
  object_def_id: string;
  runs: IIndexJobRun[];
}

// ==========================================
// Types - Error Sample
// ==========================================

export type ErrorCategory = 'SEMANTIC' | 'AI_INFERENCE' | 'MEDIA_IO' | 'SYSTEM';

export interface IIndexErrorSample {
  id: string;
  job_run_id: string;
  raw_row_id: string;
  error_category: ErrorCategory;
  error_message: string;
  stack_trace: string | null;
  created_at: string | null;
}

// ==========================================
// Types - Reindex Response
// ==========================================

export interface IReindexResponse {
  message: string;
  object_def_id: string;
  mapping_id: string;
}

// ==========================================
// API Functions
// ==========================================

/**
 * Get system health summary with all object statuses.
 */
export const fetchHealthSummary = async (): Promise<ISystemHealthSummary> => {
  const response = await v3Client.get('/health/summary');
  return response.data;
};

/**
 * Get historical job runs for an object type.
 */
export const fetchObjectHistory = async (
  objectDefId: string,
  days: number = 7,
  limit: number = 50
): Promise<IJobHistoryResponse> => {
  const response = await v3Client.get(`/health/objects/${objectDefId}/history`, {
    params: { days, limit }
  });
  return response.data;
};

/**
 * Get details of a specific job run.
 */
export const fetchJobRun = async (runId: string): Promise<IIndexJobRun> => {
  const response = await v3Client.get(`/health/jobs/${runId}`);
  return response.data;
};

/**
 * Get error samples for a specific job run.
 */
export const fetchJobErrors = async (
  runId: string,
  category?: ErrorCategory,
  limit: number = 100
): Promise<IIndexErrorSample[]> => {
  const params: Record<string, any> = { limit };
  if (category) {
    params.category = category;
  }
  const response = await v3Client.get(`/health/jobs/${runId}/errors`, { params });
  return response.data || [];
};

/**
 * Trigger immediate re-indexing for an object type.
 */
export const triggerReindex = async (objectDefId: string): Promise<IReindexResponse> => {
  const response = await v3Client.post(`/health/objects/${objectDefId}/reindex`);
  return response.data;
};

// ==========================================
// Helper Functions
// ==========================================

/**
 * Format lag time to human-readable string.
 */
export const formatLag = (lagSeconds: number | null): string => {
  if (lagSeconds === null) return 'N/A';
  
  if (lagSeconds < 60) return `${lagSeconds}s ago`;
  if (lagSeconds < 3600) return `${Math.floor(lagSeconds / 60)}m ago`;
  if (lagSeconds < 86400) return `${Math.floor(lagSeconds / 3600)}h ago`;
  return `${Math.floor(lagSeconds / 86400)}d ago`;
};

/**
 * Get status color for Ant Design Badge.
 */
export const getStatusColor = (status: HealthStatus): string => {
  switch (status) {
    case 'HEALTHY': return 'success';
    case 'DEGRADED': return 'warning';
    case 'FAILED': return 'error';
    default: return 'default';
  }
};

/**
 * Get error category color.
 */
export const getErrorCategoryColor = (category: ErrorCategory): string => {
  switch (category) {
    case 'SEMANTIC': return 'orange';
    case 'AI_INFERENCE': return 'purple';
    case 'MEDIA_IO': return 'red';
    case 'SYSTEM': return 'volcano';
    default: return 'default';
  }
};

File: backend/app/api/v3/health.py
"""
Health API endpoints for MDP Platform V3.1
Index Health monitoring and observability.
"""
from typing import List, Optional
from datetime import datetime, timedelta
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from sqlmodel import Session, select, func
from sqlalchemy import desc

from app.core.db import get_session
from app.core.logger import logger
from app.models.observability import (
    IndexJobRun,
    IndexErrorSample,
    IndexJobRunRead,
    IndexErrorSampleRead,
    ObjectHealthSummary,
    SystemHealthSummary,
    JobHistoryResponse,
)

router = APIRouter(prefix="/health", tags=["Health"])


# ==========================================
# Health Summary Endpoints
# ==========================================

@router.get("/summary", response_model=SystemHealthSummary)
def get_health_summary(session: Session = Depends(get_session)):
    """
    Get overall system health summary.
    Aggregates latest job runs for all object types.
    """
    # Get latest run for each object_def_id
    subquery = (
        select(
            IndexJobRun.object_def_id,
            func.max(IndexJobRun.start_time).label("latest_time")
        )
        .group_by(IndexJobRun.object_def_id)
        .subquery()
    )
    
    stmt = (
        select(IndexJobRun)
        .join(
            subquery,
            (IndexJobRun.object_def_id == subquery.c.object_def_id) &
            (IndexJobRun.start_time == subquery.c.latest_time)
        )
        .order_by(desc(IndexJobRun.start_time))
    )
    
    latest_runs = session.exec(stmt).all()
    
    # Build object summaries
    objects: List[ObjectHealthSummary] = []
    total_rows_processed = 0
    total_rows_indexed = 0
    total_ai_latency = 0
    ai_latency_count = 0
    total_corrupt_files = 0
    
    for run in latest_runs:
        # Calculate status
        if run.status == "FAILED":
            status = "FAILED"
        elif run.rows_indexed < run.rows_processed:
            status = "DEGRADED"
        else:
            status = "HEALTHY"
        
        # Calculate lag
        lag_seconds = None
        if run.end_time:
            lag_seconds = int((datetime.utcnow() - run.end_time).total_seconds())
        
        # Calculate success rate
        success_rate = 100.0
        if run.rows_processed > 0:
            success_rate = round((run.rows_indexed / run.rows_processed) * 100, 2)
        
        # Extract metrics
        metrics = run.metrics_json or {}
        
        # Accumulate for system totals
        total_rows_processed += run.rows_processed
        total_rows_indexed += run.rows_indexed
        
        if "ai_latency_avg_ms" in metrics:
            total_ai_latency += metrics["ai_latency_avg_ms"]
            ai_latency_count += 1
        
        if "corrupt_media_files" in metrics:
            total_corrupt_files += metrics["corrupt_media_files"]
        
        objects.append(ObjectHealthSummary(
            object_def_id=run.object_def_id,
            object_name=None,  # Can be joined with object type table
            status=status,
            last_run_time=run.end_time or run.start_time,
            lag_seconds=lag_seconds,
            rows_processed=run.rows_processed,
            rows_indexed=run.rows_indexed,
            success_rate=success_rate,
            metrics=metrics
        ))
    
    # Calculate system-level stats
    healthy_count = sum(1 for o in objects if o.status == "HEALTHY")
    degraded_count = sum(1 for o in objects if o.status == "DEGRADED")
    failed_count = sum(1 for o in objects if o.status == "FAILED")
    
    overall_success_rate = 100.0
    if total_rows_processed > 0:
        overall_success_rate = round((total_rows_indexed / total_rows_processed) * 100, 2)
    
    avg_ai_latency = None
    if ai_latency_count > 0:
        avg_ai_latency = round(total_ai_latency / ai_latency_count, 2)
    
    return SystemHealthSummary(
        total_objects=len(objects),
        healthy_count=healthy_count,
        degraded_count=degraded_count,
        failed_count=failed_count,
        overall_success_rate=overall_success_rate,
        avg_ai_latency_ms=avg_ai_latency,
        total_corrupt_files=total_corrupt_files,
        objects=objects
    )


# ==========================================
# Object History Endpoints
# ==========================================

@router.get("/objects/{object_def_id}/history", response_model=JobHistoryResponse)
def get_object_history(
    object_def_id: str,
    days: int = 7,
    limit: int = 50,
    session: Session = Depends(get_session)
):
    """
    Get historical job runs for an object type.
    Used for trend charts (e.g., AI latency over time).
    """
    since = datetime.utcnow() - timedelta(days=days)
    
    stmt = (
        select(IndexJobRun)
        .where(IndexJobRun.object_def_id == object_def_id)
        .where(IndexJobRun.start_time >= since)
        .order_by(desc(IndexJobRun.start_time))
        .limit(limit)
    )
    
    runs = session.exec(stmt).all()
    
    return JobHistoryResponse(
        object_def_id=object_def_id,
        runs=[IndexJobRunRead.model_validate(r) for r in runs]
    )


# ==========================================
# Error Sample Endpoints
# ==========================================

@router.get("/jobs/{run_id}/errors", response_model=List[IndexErrorSampleRead])
def get_job_errors(
    run_id: str,
    category: Optional[str] = None,
    limit: int = 100,
    session: Session = Depends(get_session)
):
    """
    Get error samples for a specific job run.
    Used for debugging failed/degraded runs.
    """
    stmt = select(IndexErrorSample).where(IndexErrorSample.job_run_id == run_id)
    
    if category:
        stmt = stmt.where(IndexErrorSample.error_category == category)
    
    stmt = stmt.order_by(desc(IndexErrorSample.created_at)).limit(limit)
    
    errors = session.exec(stmt).all()
    return [IndexErrorSampleRead.model_validate(e) for e in errors]


# ==========================================
# Reindex Endpoint
# ==========================================

@router.post("/objects/{object_def_id}/reindex")
def trigger_reindex(
    object_def_id: str,
    background_tasks: BackgroundTasks,
    session: Session = Depends(get_session)
):
    """
    Trigger immediate re-indexing for an object type.
    Finds the latest mapping and starts a new job.
    """
    from app.models.context import ObjectMappingDef
    from app.engine.indexing_worker import run_indexing_job
    
    # Find latest published mapping for this object
    stmt = (
        select(ObjectMappingDef)
        .where(ObjectMappingDef.object_def_id == object_def_id)
        .where(ObjectMappingDef.status == "PUBLISHED")
        .order_by(desc(ObjectMappingDef.updated_at))
        .limit(1)
    )
    
    mapping = session.exec(stmt).first()
    
    if not mapping:
        raise HTTPException(
            status_code=404,
            detail=f"No published mapping found for object: {object_def_id}"
        )
    
    # Trigger background indexing
    background_tasks.add_task(run_indexing_job, mapping.id)
    
    logger.info(f"[Health] Triggered reindex for object: {object_def_id}, mapping: {mapping.id}")
    
    return {
        "message": "Reindex job triggered",
        "object_def_id": object_def_id,
        "mapping_id": mapping.id
    }


# ==========================================
# Job Run Details
# ==========================================

@router.get("/jobs/{run_id}", response_model=IndexJobRunRead)
def get_job_run(
    run_id: str,
    session: Session = Depends(get_session)
):
    """Get details of a specific job run."""
    stmt = select(IndexJobRun).where(IndexJobRun.id == run_id)
    run = session.exec(stmt).first()
    
    if not run:
        raise HTTPException(status_code=404, detail="Job run not found")
    
    return IndexJobRunRead.model_validate(run)
